frappe/public/js/frappe/form/grid_row.js

```diff
@@ -393,11 +393,16 @@ export default class GridRow {
 		// sync get_query
 		field.get_query = this.grid.get_field(df.fieldname).get_query;
 
-		var field_on_change_function = field.df.onchange;
-		field.df.onchange = function(e) {
-			field_on_change_function && field_on_change_function(e);
-			me.grid.grid_rows[this.doc.idx - 1].refresh_field(field.df.fieldname);
-		};
+		if (!field.df.onchange_modified) {
+			var field_on_change_function = field.df.onchange;
+			field.df.onchange = function(e) {
+				field_on_change_function && field_on_change_function(e);
+				me.grid.grid_rows[this.doc.idx - 1].refresh_field(this.df.fieldname);
+			};
+
+			field.df.onchange_modified = true;
+		}
```

Comment:
```suggestion
		var field_on_change_function = field.df.onchange;
		field.df.onchange = function(e) {
			field_on_change_function && field_on_change_function(e);
			me.grid.grid_rows[field.doc.idx - 1].refresh_field(field.df.fieldname);
		};
```

Just replacing `this.doc.idx` with `field.doc.idx` as done in https://github.com/frappe/frappe/pull/11399 should fix the issue!

Reply:
It would be the best way though, @prasanna.

Let me illustrate why:

first call: works as expected.
second call: calls first function, then refreshes field again.
third call: calls second function, then refreshes field again (three times for the same field!)

I hope you get the pattern here..